ocaml/xapi/repository_helpers.ml

```diff
@@ -157,37 +157,74 @@ let create_repository_record ~__context ~name_label ~name_description
+
+  let host_exn t =
+    try host_exn t
+    with Invalid_argument _ -> (
+      match Ipaddr.of_string (Domain_name.to_string t) with
+      | Ok ip ->
+          Ipaddr.to_domain_name ip
+      | Error _ ->
+          invalid_arg "invalid host name"
+    )
+end
+
 let assert_url_is_valid ~url =
   try
     let uri = Uri.of_string url in
-    match Uri.scheme uri with
-    | Some "http" | Some "https" -> (
-      match Uri.host uri with
-      | Some h -> (
-        match !Xapi_globs.repository_domain_name_allowlist with
-        | [] ->
+    match Uri.(scheme uri, host uri) with
+    | Some ("http" | "https"), Some h -> (
+        let subdomain =
+          DomainNameIncludeIP.(host_exn (of_string h |> Result.get_ok))
+        in
+        let invalids, valids =
+          List.partition_map
+            (fun d ->
+              match Domain_name.of_string d with
+              | Ok dom ->
+                  Right dom
+              | Error _ ->
+                  Left d
+            )
+            !Xapi_globs.repository_domain_name_allowlist
+        in
+        let hostname_allowed =
+          List.exists (fun d ->
+              DomainNameIncludeIP.(is_subdomain ~subdomain ~domain:(host_exn d))
+          )
+        in
+        match (valids, invalids) with
+        | [], [] ->
             ()
-        | l -> (
-          match
-            List.exists (fun d -> Astring.String.is_suffix ~affix:("." ^ d) h) l
-          with
-          | true ->
-              ()
-          | false ->
-              let msg = "host is not in allowlist" in
-              raise Api_errors.(Server_error (internal_error, [msg]))
-        )
+        | valids, [] when not (hostname_allowed valids) ->
+            let msg = "host is not in allowlist" in
+            raise Api_errors.(Server_error (internal_error, [msg]))
+        | _, [] ->
+            ()
+        | _ ->
+            error "Invalid domains: %s in repository-domain-name-allowlist."
+              (String.concat "," invalids) ;
+            raise
+              Api_errors.(
+                Server_error (invalid_repository_domain_allowlist, invalids)
+              )
       )
-      | None ->
-          raise
-            Api_errors.(Server_error (internal_error, ["invalid host in url"]))
-    )
+    | _, None ->
+        raise
+          Api_errors.(Server_error (internal_error, ["invalid host in url"]))
     | _ ->
         raise
           Api_errors.(Server_error (internal_error, ["invalid scheme in url"]))
-  with e ->
-    error "Invalid url %s: %s" url (ExnHelper.string_of_exn e) ;
-    raise Api_errors.(Server_error (invalid_base_url, [url]))
+  with e -> (
+    match ExnHelper.error_of_exn e with
+    | error, l when error = Api_errors.invalid_repository_domain_allowlist ->
+        raise Api_errors.(Server_error (error, l))
+    | _ ->
+        error "Invalid url %s: %s" url (ExnHelper.string_of_exn e) ;
+        raise Api_errors.(Server_error (invalid_base_url, [url]))
+  )
```

Comment:
I'm not sure why error_of_exn is used, there's no need for it or a nested match:
```suggestion
  with
  | Api_errors.Server_error (err, _) as e
    when err = Api_errors.invalid_repository_domain_allowlist ->
      raise e
  | e ->
      error "Invalid url %s: %s" url (ExnHelper.string_of_exn e) ;
      raise Api_errors.(Server_error (invalid_base_url, [url]))
```

Reply:
Yeah, I didn't know it could directly match before.